pilot/pkg/security/authz/builder/builder.go

```diff
@@ -39,70 +42,108 @@ var (
+	namespace := in.Node.ConfigNamespace
+	workload := in.Node.Metadata.Labels
+	policies := in.Push.AuthzPolicies.ListAuthorizationPolicies(namespace, labels.Collection{workload})
+
+	if option.IsCustomBuilder {
+		if len(policies.Custom) == 0 {
+			authzLog.Debugf("No CUSTOM policy for workload %v in %s", workload, namespace)
+			return nil
+		}
+		return &Builder{
+			customPolicies:    policies.Custom,
+			extensions:        in.Push.Mesh.ExtensionProviders,
+			trustDomainBundle: trustDomainBundle,
+			option:            option,
+		}
+	}
+
+	if len(policies.Deny) == 0 && len(policies.Allow) == 0 && len(policies.Audit) == 0 {
+		authzLog.Debugf("No ALLOW/DENY/AUDIT policy for workload %v in %s", workload, namespace)
 		return nil
 	}
 	return &Builder{
+		denyPolicies:      policies.Deny,
+		allowPolicies:     policies.Allow,
+		auditPolicies:     policies.Audit,
 		trustDomainBundle: trustDomainBundle,
-		denyPolicies:      denyPolicies,
-		allowPolicies:     allowPolicies,
-		auditPolicies:     auditPolicies,
 		option:            option,
 	}
 }
 
-// BuildHTTP returns the RBAC HTTP filters built from the authorization policy.
+// BuildHTTP returns the HTTP filters built from the authorization policy.
 func (b Builder) BuildHTTP() []*httppb.HttpFilter {
-	var filters []*httppb.HttpFilter
+	if b.option.IsCustomBuilder {
+		// Use the DENY action so that a HTTP rule is properly handled when generating for TCP filter chain.
+		if configs := b.build(b.customPolicies, rbacpb.RBAC_DENY, false); configs != nil {
+			return configs.http
+		}
+		return nil
+	}
 
-	if auditConfig := build(b.auditPolicies, b.trustDomainBundle, rbacpb.RBAC_LOG, false, b.option); auditConfig != nil {
-		filters = append(filters, createHTTPFilter(auditConfig))
+	var filters []*httppb.HttpFilter
+	if configs := b.build(b.auditPolicies, rbacpb.RBAC_LOG, false); configs != nil {
+		filters = append(filters, configs.http...)
 	}
-	if denyConfig := build(b.denyPolicies, b.trustDomainBundle, rbacpb.RBAC_DENY, false, b.option); denyConfig != nil {
-		filters = append(filters, createHTTPFilter(denyConfig))
+	if configs := b.build(b.denyPolicies, rbacpb.RBAC_DENY, false); configs != nil {
+		filters = append(filters, configs.http...)
 	}
-	if allowConfig := build(b.allowPolicies, b.trustDomainBundle, rbacpb.RBAC_ALLOW, false, b.option); allowConfig != nil {
-		filters = append(filters, createHTTPFilter(allowConfig))
+	if configs := b.build(b.allowPolicies, rbacpb.RBAC_ALLOW, false); configs != nil {
+		filters = append(filters, configs.http...)
 	}
-
 	return filters
 }
 
-// BuildTCP returns the RBAC TCP filters built from the authorization policy.
+// BuildTCP returns the TCP filters built from the authorization policy.
 func (b Builder) BuildTCP() []*tcppb.Filter {
-	var filters []*tcppb.Filter
+	if b.option.IsCustomBuilder {
+		if configs := b.build(b.customPolicies, rbacpb.RBAC_DENY, true); configs != nil {
```

Comment:
Why is custom builder using DENY filter?

Reply:
DENY will cause an invalid config (e.g. found HTTP attribute in TCP filter chain) to always trigger the action (calling ext_authz), it doesn't matter in runtime as it's shadow mode (only evaluate but not enforce).

Reply:
I see. Thanks.